pkg/controllers/util/federatedclient/podinformer_test.go

```diff
@@ -0,0 +1,236 @@
+	"k8s.io/client-go/kubernetes"
+	kubeclient "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/fake"
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/kubewharf/kubeadmiral/pkg/controllers/util"
+)
+
+func setupFakeClient(numPods int) (pods []runtime.Object, client kubernetes.Interface) {
+	pods = make([]runtime.Object, numPods)
+	for i := 1; i <= numPods; i++ {
+		pods[i-1] = newPod(i)
+	}
+	return pods, fake.NewSimpleClientset(pods...)
+}
+
+func newPod(id int) *corev1.Pod {
+	return &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   fmt.Sprintf("test-pod-%d", id),
+			Labels: map[string]string{"foo": "bar"},
+		},
+		Spec: corev1.PodSpec{
+			Overhead: corev1.ResourceList{
+				corev1.ResourceCPU: resource.MustParse("1m"),
+			},
+			InitContainers: []corev1.Container{
+				{
+					Name: "initcontainer",
+					Resources: corev1.ResourceRequirements{
+						Limits: corev1.ResourceList{
+							corev1.ResourceCPU: resource.MustParse("1m"),
+						},
+						Requests: corev1.ResourceList{
+							corev1.ResourceCPU: resource.MustParse("1m"),
+						},
+					},
+				},
+			},
+			Containers: []corev1.Container{
+				{
+					Name: "container",
+					Resources: corev1.ResourceRequirements{
+						Limits: corev1.ResourceList{
+							corev1.ResourceCPU: resource.MustParse("10m"),
+						},
+						Requests: corev1.ResourceList{
+							corev1.ResourceCPU: resource.MustParse("10m"),
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func Test_federatedClientFactory_podInformer(t *testing.T) {
+	type fields struct {
+		kubeClients         map[string]kubeclient.Interface
+		informers           map[string]kubeinformer.SharedInformerFactory
+		availablePodListers int32
+	}
+	tests := []struct {
+		name              string
+		clusters          int
+		podNum            int
+		expectCacheSynced bool
+		fields            fields
+	}{
+		{
+			name:              "4 clusters, limit 2",
+			clusters:          4,
+			podNum:            10000,
+			expectCacheSynced: true,
+			fields: fields{
+				kubeClients:         map[string]kubeclient.Interface{},
+				informers:           map[string]kubeinformer.SharedInformerFactory{},
+				availablePodListers: 2,
+			},
+		},
+		{
+			name:              "4 clusters, limit 0",
+			clusters:          4,
+			podNum:            10000,
+			expectCacheSynced: false,
+			fields: fields{
+				kubeClients:         map[string]kubeclient.Interface{},
+				informers:           map[string]kubeinformer.SharedInformerFactory{},
+				availablePodListers: 0,
+			},
+		},
+	}
+	handlerFn := func(obj interface{}, podName string, ch chan int) error {
+		pod, ok := obj.(*corev1.Pod)
+		if !ok {
+			return fmt.Errorf("expected to handle pod, but got: %v", obj)
+		}
+		if pod.Name == podName {
+			ch <- 1
+		}
+		if pod.Labels != nil {
+			return fmt.Errorf("expect no label for pod, but got labels: %v", pod.Labels)
+		}
+		return nil
+	}
+
+	for _, tt := range tests {
+		for i := 0; i < tt.clusters; i++ {
```

Comment:
This initialisation block should be inside `t.Run` as well, so any panics can be linked to the subtest.